### PR TITLE
TTL for cleaning up Pods after they finish

### PR DESCRIPTION
**Pull request by alaypatel07**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**: [enhancements/issues/592](https://github.com/kubernetes/enhancements/issues/592)

**Which issue(s) this PR fixes**:

Fixes #
For [enhancements/issues/592](https://github.com/kubernetes/enhancements/issues/592)

**Special notes for your reviewer**: @kubernetes/sig-apps-pr-reviews

**Does this PR introduce a user-facing change?**:

```release-note
Add a TTL mechanism to clean up Pods after they finish.
```

### CODE REVIEW

File: pkg/controller/ttlafterfinished/ttlafterfinished_controller.go [hunk 1]
```diff
@@ -20,92 +20,107 @@ import (
...
+	// Resources that the controller will check its TTL and attempt to delete when the TTL expires.
 	queue workqueue.RateLimitingInterface
 
 	// The clock for tracking time
 	clock clock.Clock
 }
 
-// New creates an instance of Controller
-func New(jobInformer batchinformers.JobInformer, client clientset.Interface) *Controller {
+// New creates an instance o Controller
```
Comment (janetkuo) [inline pkg/controller/ttlafterfinished/ttlafterfinished_controller.go]: typo?

File: cmd/kube-controller-manager/app/core.go [hunk 2]
```diff
@@ -481,9 +482,15 @@ func startTTLAfterFinishedController(ctx ControllerContext) (http.Handler, bool,
...
-	).Run(int(ctx.ComponentConfig.TTLAfterFinishedController.ConcurrentTTLSyncs), ctx.Stop)
+		ctx.InformerFactory,
+		schema.GroupVersionResource{Group: "batch", Version: "v1", Resource: "jobs"},
+		dynamic.NewForConfigOrDie(ctx.ClientBuilder.ConfigOrDie("ttl-jobs-after-finished-controller")),
+		ctx.ClientBuilder.ClientOrDie("ttl-jobs-after-finished-controller")).Run(int(ctx.ComponentConfig.TTLAfterFinishedController.ConcurrentTTLSyncs), ctx.Stop)
+	go ttlafterfinished.New(
+		ctx.InformerFactory,
+		schema.GroupVersionResource{Version: "v1", Resource: "pods"},
+		dynamic.NewForConfigOrDie(ctx.ClientBuilder.ConfigOrDie("ttl-pods-after-finished-controller")),
+		ctx.ClientBuilder.ClientOrDie("ttl-pods-after-finished-controller")).Run(int(ctx.ComponentConfig.TTLAfterFinishedController.ConcurrentTTLSyncs), ctx.Stop)
```
Comment (janetkuo) [inline cmd/kube-controller-manager/app/core.go]: We only need to start a single instance of TTL controller that handles both Jobs and Pods.

File: pkg/controller/ttlafterfinished/ttlafterfinished_controller.go (hunk 1 above)
Comment (alaypatel07) [reply pkg/controller/ttlafterfinished/ttlafterfinished_controller.go]: yeah, I'll fix it. Thanks for spotting!

File: pkg/controller/ttlafterfinished/ttlafterfinished_controller.go [hunk 3]
```diff
@@ -20,92 +20,107 @@ import (
...
 	defer tc.queue.ShutDown()
 
-	klog.Infof("Starting TTL after finished controller")
+	klog.Infof("Starting TTL after finished controller for %s", tc.resource.String())
 	defer klog.Infof("Shutting down TTL after finished controller")
 
-	if !controller.WaitForCacheSync("TTL after finished", stopCh, tc.jListerSynced) {
+	if !controller.WaitForCacheSync("TTL after finished", stopCh, tc.rListerSynced) {
 		return
 	}
```
Comment (janetkuo) [inline pkg/controller/ttlafterfinished/ttlafterfinished_controller.go]: We could just run a dedicated worker for Pods, and a dedicated queue/informer/lister for Pods too. 

We could use something more generic like what [garbage collector](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/garbagecollector.go) uses, but that seems to be premature optimization. We can move to use that if we're supporting custom resources.

Comment (alaypatel07) [reply pkg/controller/ttlafterfinished/ttlafterfinished_controller.go]: @janetkuo Thanks. I’ll refactor the code with dedicated queue/informer/lister for pods and jobs

Out of curiosity, can I ask why not use two controllers for jobs and pods if we are using dedicated queues/informers/lister(which I think is the crux of a controller) anyways? Am I missing anything?

**Comment by xmudrii** [comment]
@alaypatel07 @janetkuo @tnozicka Hello! I'm the bug triage lead for the 1.16 release cycle and considering this PR is tagged for 1.16, I'd like to check its status. The code freeze is starting on August 29th EOD PST (about 5 days from now) and once the code freeze starts, only release-blocking PRs will be considered to be merged. Is this PR still planned for the 1.16 release?

File: pkg/controller/ttlafterfinished/ttlafterfinished_controller.go [hunk 4]
```diff
@@ -179,93 +195,109 @@ func (tc *Controller) handleErr(err error, key interface{}) {
...
-	tc.enqueueAfter(job, *t)
+	tc.enqueueAfter(u, *t)
 	return false, nil
 }
 
-// needsCleanup checks whether a Job has finished and has a TTL set.
-func needsCleanup(j *batch.Job) bool {
+// needsCleanup checks whether a resource has finished and has a TTL set.
+func needsCleanup(obj interface{}) bool {
+	if job, ok := obj.(*batch.Job); ok {
```
Comment (tnozicka) [inline pkg/controller/ttlafterfinished/ttlafterfinished_controller.go]: do a switch and hard error / panic if none of the types is matched

File: cmd/kube-controller-manager/app/core.go (hunk 2 above)
Comment (tnozicka) [reply cmd/kube-controller-manager/app/core.go]: I agree, start one instance of the controller, 2 informers, 1 queue and then switch based on type where the flow would differ. We can go to dynamic informers and unstructured objects checking if we determine we need it for CRDs later.

**Comment by janetkuo** [comment]
@xmudrii yes this PR is still planned for 1.16.

**Comment by ipochi** [comment]
@tnozicka @janetkuo 

@alaypatel07 Hey I created a PR #82082 for the same
Would you be able to get your PR ready for merging. I will get reviewed and close without merging  if you can.

**Comment by alaypatel07** [comment]
@ipochi Thanks for the heads up. I am trying to finish it up today. Hopefully, that would be enough time for merging.

**Comment by liggitt** [comment]
/milestone v1.17

this has missed code freeze

**Comment by alejandrox1** [comment]
@alaypatel07 @janetkuo @enisoc @tnozicka any updates on this?
/remove-area e2e-test-framework
/remove-sig testing

/remove-area cloudprovider
/remove-area code-generation
/remove-area conformance
/remove-area dependency
/remove-area ipvs
/remove-area kubeadm
/remove-area kubectl
/remove-area release-eng
/remove-area kubelet
/remove-area test
/remove-sig auth
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig node
/remove-sig release
/remove-sig scheduling
/remove-sig storage

**Comment by josiahbjorgaard** [comment]
Bug triage for 1.17 here. Code freeze is Nov. 18th. Will this PR be merged by then? If not, can we move it out of the milestone?

**Comment by alaypatel07** [comment]
closing because another PR is tracking similar work

**Pull request closed without merging**